_posts/2021-09-07-what-javascript-tests-could-learn-from-rspec.md

```diff
@@ -0,0 +1,227 @@
+---
+layout: epic
+title: "What JavaScript Tests Could Learn From RSpec"
+date: 2021-09-10
+categories: [testing, rspec, jest]
+author: steve-hicks
+---
+
+TODO: rel="canonical" to the original on my blog
+
+When I started at Artsy a few years ago, I'd never written a line of Ruby. I feel at home with JavaScript — it's
+been my buddy since I started my career over 20 years ago. I've written enough tests in JavaScript that I sometimes
+feel like I can write them in my sleep (as long as they don't involve async React events 😅).
+
+Most of the code I write at Artsy is still JavaScript, but now I write some Ruby code too, and I've written enough
+RSpec tests that I'm starting to form opinions about what I think they should look like.
+
+My most recent work has been JavaScript again. I've been writing Jest tests against one of our React apps. But
+rather than reaching for the testing patterns I'd become accustomed to over my years of JavaScripting, I'm finding
+that something's missing in my Jest tests! My experiences with RSpec have me longing for two features in Jest:
+
+<!-- more -->
+
+1. `let` blocks
+2. `context` blocks
+
+## 1. `let` blocks
+
+[`let` blocks](https://relishapp.com/rspec/rspec-core/v/2-11/docs/helper-methods/let-and-let) are used in an RSpec
+test to set things up for your test scenario.
+
+Here's a test for a `Counter` class, verifying that when I call the `increment` method on an instance, its stored
+value becomes `1`.
+
+```rb
+describe "counter" do
+
+  let(:counter) { Counter.new }
+
+  describe "increment" do
+    it "increments by 1" do
+      counter.increment
+
+      counter.value.should eq(1)
+    end
+  end
+end
+```
+
+If you're new to Ruby, the only line that doesn't translate almost directly to a similar JavaScript expression is
+the `let` statement.
+
+The `let` statement in RSpec creates a thing named `counter`, which is evaluated to a new instance of the `Counter`
+class. There are a few important things to note about `let` blocks:
+
+1. They're evaluated lazily, so that thing doesn't actually get created until I reference it.
```

Comment:
```suggestion
1. They're evaluated lazily by default, so that thing doesn't actually get created until I reference it. If you want to eager load you can use `let!` to do that.
```

Reply:
I thought about clarifying that, but thought it might be an unnecessary distraction. I like `by default` - I I'll add that, and think about the `let!` clarification.

Reply:
I added `by default`! left out the `let!` clarification because I didn't think it was necessary. Thanks for the suggestions!